File: 2025-11/2025-11-10/extensionExtractor.js
function getExtension(filename) {

  return /\.(?<extension>[^\.]*)$/g
    .exec(filename)
    ?.groups
    ?.extension
    || "none";
}

console.log(getExtension("document.txt"));
console.log(getExtension("README"));
console.log(getExtension("image.PNG"));
console.log(getExtension(".gitignore"));
console.log(getExtension("archive.tar.gz"));
console.log(getExtension("final.draft."));

// --- TEST SUITE ---

const testsText = `
1. getExtension("document.txt") should return "txt".
2. getExtension("README") should return "none".
3. getExtension("image.PNG") should return "PNG".
4. getExtension(".gitignore") should return "gitignore".
5. getExtension("archive.tar.gz") should return "gz".
6. getExtension("final.draft.") should return "none".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);

File: 11 November/11-10/extension_extractor.py
import re

def get_extension(filename):
    match = re.search(r"\.(?P<extension>[^\.]*)$", filename)
    
    if match:
        extension = match.groupdict()['extension']
        return extension or "none"

    else:
        return "none"

print(get_extension("document.txt"))
print(get_extension("README"))
print(get_extension("image.PNG"))
print(get_extension(".gitignore"))
print(get_extension("archive.tar.gz"))
print(get_extension("final.draft."))

# --- TEST SUITE ---

tests_text = '''
1. get_extension("document.txt") should return "txt".
2. get_extension("README") should return "none".
3. get_extension("image.PNG") should return "PNG".
4. get_extension(".gitignore") should return "gitignore".
5. get_extension("archive.tar.gz") should return "gz".
6. get_extension("final.draft.") should return "none".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)